core/inc/knoting/engine.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <array>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "Window.h"
+#include "subsystem.h"
+
+namespace knot {
+class Engine {
+   public:
+    Engine();
+    ~Engine();
+
+    void update_modules();
+
+    std::shared_ptr<knot::Window> get_window_module() { return m_windowModule; }
+
+   private:
+    int m_windowWidth = 1024;
+    int m_windowHeight = 768;
+    std::string m_windowTitle = "hello knotting!";
+
+   private:
+    std::vector<std::shared_ptr<knot::Subsystem>> m_engineModules;
```

Comment:
Since this is the main place where all our subsystems will lay, are `shared_ptr`s necessary?  
Because the only reason a subsystem would be freed is that it no longer is part of the engine, in which case we want it to be fully dropped.

If `shared_ptr`s are necessary, consider returning a `weak_ptr` inside any `get_<system>_module`.
If not, let's consider using a `unique_ptr` in here, and either returning a reference in the `get_<system>_module` functions or even a raw pointer.

Reply:
I think weak_ptr for a get_<subsystem>_module is a good idea the object using the reference do not require ownership